framework/include/actions/ActionWarehouse.h

```diff
@@ -101,6 +101,64 @@ class ActionWarehouse :
   const std::vector<Action *> & getActionsByName(const std::string & task) const;
 
   /**
+   * Retrieve an action with its name and the desired type.
+   * @param name The action name.
+   */
+  template <class T>
+  const T & getAction(const std::string & name)
+    {
+      T* p = NULL;
+      for (std::vector<std::string>::iterator it = _ordered_names.begin(); it != _ordered_names.end(); ++it)
+      {
+        std::string task = *it;
+        for (ActionIterator act_iter = actionBlocksWithActionBegin(task);
+             act_iter != actionBlocksWithActionEnd(task);
+             ++act_iter)
+        {
+          if ((*act_iter)->name() == name)
+          {
+            p = dynamic_cast<T*>(*act_iter);
+            if (!p)
+              mooseError("The action with name being "+name+" is not in the desired type");
+            else
+              break;
+          }
+        }
+      }
+      if (!p)
+        mooseError("Action with name being "+name+" does not exist");
+      return *p;
+    }
+
+  /**
+   * Retrieve all actions in a specific type.
+   */
+  template <class T>
+  std::vector<const T *> getActions()
```

Comment:
`WareHouse` only has `all()` but not `get(const std::string & name)`. `ActionWarehouse` has one called `getActionsByName()` but it is indeed getting actions by a task. If I use this getter, I do not need to do dynamic cast by myself. I have repeated codes on this getter, which is why I want to generalize it into MOOSE. I cannot find a desired place to hold this code in my user-code. I can put it under user-app, but then whenever I use `_app`, I have to dynamic cast to the user-app, which makes the code ugly.